runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,573 @@
+          api.nvim_buf_add_highlight(
+            bufnr,
+            state.namespace,
+            '@' .. token.type,
+            token.line,
+            token.start_col,
+            token.end_col
+          )
+          token.extmark_added = true
+        end
+
+        --TODO(jdrouhard): do something with the modifiers
+      end
+    end
+  end
+end
+
+local M = {}
+
+--- Start the semantic token highlighting engine for the given buffer with the
+--- given client.
+---
+--- NOTE: This is currently called automatically by |buf_attach_client|. To
+--- opt-out of semantic highlighting with a server that supports it, you can
+--- delete the semanticTokensProvider table from the {server_capabilities} of
+--- your client in your |LspAttach| callback or your configuration's
+--- `on_attach` callback.
+---
+--- <pre>
+--- client.server_capabilities.semanticTokensProvider = nil
+--- </pre>
+---
+---@param bufnr number
+---@param client_id number
+function M.start(bufnr, client_id)
+  vim.validate({
+    bufnr = { bufnr, 'n', false },
+    client_id = { client_id, 'n', false }
+  })
+
+  local client = vim.lsp.get_client_by_id(client_id)
+  if not client then
+    vim.notify("No LSP server configured with client_id " .. client_id)
+    return
+  end
+  if not vim.tbl_get(client.server_capabilities, 'semanticTokensProvider', 'full') then
+    vim.notify('Language server does not support semantic tokens')
+    return
+  end
+
+  active_buffers[bufnr] = true
+  local state = state_by_client[client_id]
+  if not state then
+    state = {
+      buffers = {},
+      namespace = api.nvim_create_namespace('vim_lsp_semantic_tokens:' .. client_id)
+    }
+    state_by_client[client_id] = state
+  end
+
+  local buf_state = {
+    current_result = {},
+    active_request = {},
+  }
+  state.buffers[bufnr] = buf_state
+
+  api.nvim_buf_attach(bufnr, false, {
+    on_lines = function()
+      -- detach if the LSP client detached from this buffer or on_detach() was
+      -- called manually
+      local s, bs = get_state(bufnr, client_id)
+      if not s or not bs or not vim.lsp.buf_is_attached(bufnr, client_id) then
+        return true
+      end
+    end,
+    on_reload = function()
+      api.nvim_buf_clear_namespace(bufnr, state.namespace, 0, -1)
+      reset_buf_state(bufnr, client, true)
+      request_all_tokens(bufnr, client_id)
+    end,
+    on_detach = function()
+      M.stop(bufnr, client_id)
+    end,
+  })
+
+  -- Add a BufWinEnter event to re-highlight buffers whose state was reset
+  -- without a document version change, such as after receiving a refresh
+  -- request from the server while the buffer was hidden
+  local augroup_str = ('vim_lsp_c_%d_b_%d_semantic_tokens'):format(client_id, bufnr)
+  local augroup = api.nvim_create_augroup(augroup_str, { clear = true })
+  api.nvim_create_autocmd({ 'BufWinEnter' }, {
+    buffer = bufnr,
+    group = augroup,
+    callback = function()
+      request_all_tokens(bufnr, client_id)
+    end,
+  })
```

Comment:
Instead of a `BufWinEnter` event, we could also just re-request tokens for all buffers for a client on a refresh request immediately. I don't really think one or the other is better. This current way is a little less chatter between nvim and the LSP server, but re-requesting all buffers immediately would be "snappier" for the user since switching back to a buffer would already have updated tokens and there'd be no delay waiting.

Reply:
I think the `BufWinEnter` autocmd is the way to go, here. Spread the rpc/server load out if there's a refresh.